ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/TransactionContractKeysValidation.scala

```diff
@@ -0,0 +1,295 @@
+    val causalKeyMonotonicity = keys.forall { key =>
+      val state = damlState(key)
+      val keyActiveAt =
+        Conversions.parseTimestamp(state.getContractKeyState.getActiveAt).toInstant
+      !keyActiveAt.isAfter(transactionEntry.ledgerEffectiveTime.toInstant)
+    }
+    if (causalKeyMonotonicity)
+      StepContinue(transactionEntry)
+    else
+      transactionCommitter.reject(
+        recordTime,
+        transactionCommitter.buildRejectionLogEntry(
+          transactionEntry,
+          RejectionReason.Inconsistent("Causal monotonicity violated"),
+        ),
+      )
+  }
+
+  private sealed abstract class KeyValidationError()
+  private case object Duplicate extends KeyValidationError
+  private case object Inconsistent extends KeyValidationError
+
+  private case class KeyValidationState(
+      activeDamlStateKeys: Set[DamlStateKey],
+      submittedDamlContractKeysToContractIds: Map[DamlContractKey, Option[RawContractId]] =
+        Map.empty,
+  ) {
+    def addSubmittedDamlContractKeyToContractIdIfEmpty(
+        submittedDamlContractKey: DamlContractKey,
+        contractId: Option[ContractId],
+    ): KeyValidationState =
+      copy(submittedDamlContractKeysToContractIds =
+        if (!submittedDamlContractKeysToContractIds.contains(submittedDamlContractKey)) {
+          submittedDamlContractKeysToContractIds + (submittedDamlContractKey -> contractId.map(
+            _.coid
+          ))
+        } else {
+          submittedDamlContractKeysToContractIds
+        }
+      )
+  }
+
+  private type KeyValidationStatus = Either[KeyValidationError, KeyValidationState]
+
+  private def validateTransactionContractKeysUniquenessAndConsistency(
+      transactionCommitter: TransactionCommitter,
+      recordTime: Option[Timestamp],
+      transactionEntry: DamlTransactionEntrySummary,
+      contractKeyDamlStateKeys: Set[DamlStateKey],
+      contractKeysToContractIds: Map[DamlContractKey, RawContractId],
+  ): StepResult[DamlTransactionEntrySummary] = {
+
+    val keysValidationOutcome = transactionEntry.transaction
+      .foldInExecutionOrder[KeyValidationStatus](
+        Right(KeyValidationState(activeDamlStateKeys = contractKeyDamlStateKeys))
+      )(
+        (keyValidationStatus, _, exerciseBeginNode) =>
+          validateNodeContractKey(
+            exerciseBeginNode,
+            contractKeysToContractIds,
+            keyValidationStatus,
+          ),
+        (keyValidationStatus, _, leafNode) =>
+          validateNodeContractKey(leafNode, contractKeysToContractIds, keyValidationStatus),
+        (accum, _, _) => accum,
+      )
+
+    keysValidationOutcome match {
+      case Right(_) =>
+        StepContinue(transactionEntry)
+      case Left(error) =>
+        val message = error match {
+          case Duplicate => "DuplicateKeys: at least one contract key is not unique"
+          case Inconsistent =>
+            "InconsistentKeys: at least one contract key has changed since the submission"
```

Comment:
In #8335 we included the keys in the inconsistent case but I'd like to leave that for a future PR because 1. it's more computation and 2. for consistency we should really do for the duplicate case as well.